src/Shape.cpp

```diff
@@ -141,19 +188,56 @@ bool Shape::intersect(const Pose3D& pose, const Shape& other) const {
             double d_v2_1 = N_2.dot(v2_1) + d_2;
             double d_v3_1 = N_2.dot(v3_1) + d_2;
 
+            if ((d_v1_1 == 0 && d_v2_1 == 0 && d_v3_1 == 0) || (d_v1_1 == 0 && d_v2_1 == 0 && d_v3_1 == 0)) {
+                if (N_1.dot(N_2) < 0) // triangles face opposite directions.
+                    continue;
+
+                // 2D triangle triangle test
+
+                // convert vectors to 2D
+                Vec2d v1_1_2d, v2_1_2d, v3_1_2d;
+                Vec2d v1_2_2d, v2_2_2d, v3_2_2d;
+
+                double maxN_1 = std::max(std::max(N_1.x, N_1.y), N_1.z);
+                int skip = (maxN_1 == N_1.y) + 2*(maxN_1==N_2.z);
+
+                project3Dto2Dvector(v1_1_2d, v1_1, skip);
+                project3Dto2Dvector(v2_1_2d, v2_1, skip);
+                project3Dto2Dvector(v3_1_2d, v3_1, skip);
+
+                project3Dto2Dvector(v1_2_2d, v1_2, skip);
+                project3Dto2Dvector(v2_2_2d, v2_2, skip);
+                project3Dto2Dvector(v3_2_2d, v3_2, skip);
+
+                if (linelineintersect(v1_1_2d, v2_1_2d, v1_2_2d, v2_2_2d)) return true;
+                if (linelineintersect(v1_1_2d, v2_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v1_1_2d, v2_1_2d, v3_2_2d, v1_2_2d)) return true;
+
+                if (linelineintersect(v2_1_2d, v3_1_2d, v1_2_2d, v2_2_2d)) return true;
+                if (linelineintersect(v2_1_2d, v3_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v2_1_2d, v3_1_2d, v3_2_2d, v1_2_2d)) return true;
+
+                if (linelineintersect(v3_1_2d, v1_1_2d, v1_2_2d, v2_2_2d)) return true;
+                if (linelineintersect(v3_1_2d, v1_1_2d, v2_2_2d, v3_2_2d)) return true;
+                if (linelineintersect(v3_1_2d, v1_1_2d, v3_2_2d, v1_2_2d)) return true;
```

Comment:
Is is very ugly. There should definitely be a better solution for this.

Reply:
Made something different for this but I don't know if that makes it better.